lib/parser/Parser.cpp

```diff
@@ -0,0 +1,1609 @@
+#include "Parser.hpp"
+
+namespace ptx_parser {
+
+std::string Node::repr() {
+  std::string output = "";
+  if (children.size() > 0) {
+    output = output + "(";
+  }
+  if (node_type == leaf) {
+    std::cout << "found leaf" << std::endl;
+    output = output + value.text;
+  }
+  if (children.size() > 0) {
+    for (Node child : children) {
+      output = output + child.repr();
+    }
+    output = output + ")";
+  }
+  return output;
+}
+
+void Parser::parse(std::vector<Token> tokens) {
+  current_token_index = 0;
+  int success = 0;
+  Node child_node;
+
+  while (current_token_index < tokens.size()) {
+
+    success = parse_statement(tokens, child_node);
+
+    if (success == 1) {
+      parse_tree.add(child_node);
+      std::cout << "SUCCESS" << std::endl;
+    } else {
+      std::cout << "ERROR" << std::endl;
+    }
+  }
+}
+
+void Parser::advance(std::vector<Token> tokens) {
+  current_token_index = current_token_index + 1;
+}
+
+int Parser::parse_statement(std::vector<Token> tokens, Node &node) {
```

Comment:
why not return a bool? then you can just do (if (success)) instead of (if success == 1)